src/malloc_interceptor.c

```diff
@@ -0,0 +1,602 @@
+    if (!g_memlog.initialized || !g_memlog.map) return;
+
+    size_t events = atomic_load_explicit(&g_memlog.index, memory_order_relaxed);
+    size_t used_bytes = g_memlog.header_bytes + events * sizeof(PeakMemEvent);
+    if (used_bytes > g_memlog.map_bytes) used_bytes = g_memlog.map_bytes;
+    
+    msync(g_memlog.map, used_bytes, MS_SYNC);
+
+    int fd_csv = open(g_memlog.csv_path, O_CREAT | O_TRUNC | O_WRONLY | O_CLOEXEC, 0644);
+    if (fd_csv < 0) {
+        fprintf(stderr, "[peak] memlog: open CSV %s failed: %s\n", g_memlog.csv_path, strerror(errno));
+    } else {
+        dprintf(fd_csv, "ts_ns,delta,current,tid,op\n");
+
+        uint8_t *base = (uint8_t *) g_memlog.map + g_memlog.header_bytes;
+        for (size_t i = 0; i < events; i++) {
+            PeakMemEvent *e = (PeakMemEvent *) (base + i * sizeof(PeakMemEvent));
+            peak_csv_emit_line(fd_csv, e);
+        }
+        close(fd_csv);
+    }
+    fprintf(stderr, "[peak] memlog CSV written: %s (events=%zu)\n", g_memlog.csv_path, events);
+
+    munmap(g_memlog.map, g_memlog.map_bytes);
+    (void) ftruncate(g_memlog.fd, (off_t) used_bytes);
+    close(g_memlog.fd);
+    unlink(g_memlog.tmp_path);
+
+    g_memlog.map = NULL;
+    g_memlog.map_bytes = 0;
+    g_memlog.capacity_events = 0;
+    g_memlog.initialized = 0;
+}
+
+/*=========================
+  Tracking table helpers
+=========================*/
+
+static void add_tracking_entry(void* ptr, size_t size, int is_mmap, char* filename, int log) {
+    if (!track_table || cleanup_in_progress) return;
+    if (!log) return;
+
+    AllocationEntry* entry = internal_malloc(sizeof(AllocationEntry));
+    if (!entry) return;
+
+    entry->ptr      = ptr;
+    entry->size     = size;
+    entry->is_mmap  = is_mmap;
+    entry->filename = filename;  // May be NULL for non-mmap allocations
+
+    pthread_mutex_lock(&track_mutex);
+    gum_metal_hash_table_insert(track_table, ptr, entry);
+    current_memory += size;
+    max_memory = current_memory > max_memory ? current_memory : max_memory;
+    pthread_mutex_unlock(&track_mutex);
+
+    if (log) peak_log_event((int64_t) size, (uint64_t) current_memory, 1);
+}
+
+static AllocationEntry* find_tracking_entry(void* ptr) {
+    if (!track_table || cleanup_in_progress) return NULL;
+
+    pthread_mutex_lock(&track_mutex);
+    AllocationEntry* entry = gum_metal_hash_table_lookup(track_table, ptr);
+    if (entry && entry->ptr == ptr) {
+        pthread_mutex_unlock(&track_mutex);
+        return entry;
+    }
+    pthread_mutex_unlock(&track_mutex);
+    return NULL;
+}
+
+static void remove_tracking_entry(void* ptr) {
+    if (!track_table || cleanup_in_progress) return;
+
+    pthread_mutex_lock(&track_mutex);
+    AllocationEntry* entry = gum_metal_hash_table_lookup(track_table, ptr);
+    if (entry) {
+        gum_metal_hash_table_remove(track_table, ptr);
+        current_memory -= entry->size;
+    }
+    pthread_mutex_unlock(&track_mutex);
+
+    if (entry->filename) {
```

Comment:
It seems the `add_tracking_entry` function always sets `filename = NULL`, right? We should just remove this block. The if branching actually have a performance hit.